### PR TITLE
Migrate to GRPC API 2.0

### PR DESCRIPTION
**Pull request by avive**
This is an initial refactoring of the CLIWallet which includes:
1. Phased out use of http-json api 1.0
2. Basic api 2.0 client.

This is not a full implementation of api 2.0 - just an initial one for review of the refactoring.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,11 +1,15 @@
...
 
 ## Overview
-A basic reference Spacemesh CLI Wallet designed to work together with a [go-spacemesh full node](https://github.com/spacemeshos/go-spacemesh).
+A basic reference console Spacemesh Wallet designed to work together with a [go-spacemesh full node](https://github.com/spacemeshos/go-spacemesh).
 
-The CLI Wallet is designed for developers who want to hack on the Spacemesh platform. For non-devs we recommend using Smapp - the Spacemesh App. [Smapp](https://github.com/spacemeshos/smapp) is available for all major desktop platforms and includes a wallet.
+The CLIWallet is designed for developers who want to hack on the Spacemesh platform. For non-devs we recommend using Smapp - the Spacemesh App. [Smapp](https://github.com/spacemeshos/smapp) is available for all major desktop platforms and includes a wallet.
 
 ## Functionality
+The wallet is a Spacemesh node api 2.0 client and implements basic wallet features. You can create a new coin account, execute transactions, check account balance and view transactions./
```
Comment (lrettig) [inline README.md:9]: The term "api 2.0" by itself is going to confuse people, especially since the API versioning currently starts with 1.x. It's probably best not to mention the version here at all, just call it the "Spacemesh API."

File: README.md [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
+
+> WARNING: CLIWallet is currently insecure as it saves private keys in cleartext on your local store. It is not yet a production-quality wallet. We plan to update the project to store private data securely in future releases.
```
Comment (lrettig) [inline README.md:11]: Add a link here to an issue on this topic?

File: README.md [hunk 3]
```diff
@@ -50,5 +54,18 @@ make dockerbuild-go
 ./cli_wallet
 ```
 
+By default, the wallet will attempt to connect to a locally running Spacemesh full node using the default node's grpc api port.
+You can configure the ports in go-sm and use CLIWallet `--grpc-server` and `--grpc-port` flags to override the defaults.
+To use the full wallet features, enable all GRPC services in your node's config file. e.g:
```
Comment (lrettig) [inline README.md:59]: Probably best not to document this here, this documentation should live under go-spacemesh and should link to it here.

File: client/mesh_service.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+		// todo: add warning, each result should be a transaction
+		if a != nil {
+			activations = append(activations, a)
+		}
+	}
+
+	return activations, resp.TotalResults, nil
+}
+
+func (c *gRPCClient) GetMeshInfo() (*common.NetInfo, error) {
```
Comment (lrettig) [inline client/mesh_service.go:81]: I see what you mean about how this could all be one request! :)

File: client/node_service.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	"errors"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/spacemeshos/CLIWallet/common"
+	apitypes "github.com/spacemeshos/api/release/go/spacemesh/v1"
+	"golang.org/x/net/context"
+)
+
+// Sanity is a basic api sanity test. It verifies that the client can connect to
+// the node service and get a response from it to an echo request.s
```
Comment (lrettig) [inline client/node_service.go:13]:
```suggestion
// the node service and get a response from it to an echo request.
```

File: main.go [hunk 6]
```diff
@@ -26,25 +24,28 @@ func (m mockClient) Transfer(from, to, amount, passphrase string) error {
...
 
 func main() {
-	serverHostPort := client.DefaultNodeHostPort
-	datadir := Getwd()
+	dataDir := getwd()
+
+	grpcServerPort := uint(client.DefaultGRPCPort)
+	grpcServer := client.DefaultGRPCServer
+
+	flag.StringVar(&grpcServer, "grpc-server", grpcServer, "The api 2.0 grpc server")
```
Comment (lrettig) [inline main.go:32]: Same comment as above, no need to reference API version here

File: repl/account_commands.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+	if err != nil {
+		log.Error("failed to get account info: %v", err)
+		return
+	}
+
+	fmt.Println(printPrefix, "Local alias:", acc.Name)
+	fmt.Println(printPrefix, "Address:", address.String())
+	fmt.Println(printPrefix, "Balance:", state.StateCurrent.Balance.Value, coinUnitName)
+	fmt.Println(printPrefix, "Nonce:", state.StateCurrent.Counter)
+	fmt.Println(printPrefix, "Projected Balance:", state.StateProjected.Balance.Value, coinUnitName)
```
Comment (lrettig) [inline repl/account_commands.go:66]: Might be nice to add a line to output here that says something like "Projected account state includes all pending transactions that haven't been added to the mesh yet."

File: repl/account_commands.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+	}
+
+	fmt.Println(printPrefix, fmt.Sprintf("Total rewards: %d", total))
+	for _, r := range rewards {
+		printReward(r)
+		fmt.Println(printPrefix, "-----")
+	}
+}
+
+func printReward(r *apitypes.Reward) {
```
Comment (lrettig) [inline repl/account_commands.go:95]: Empty function??

File: repl/debug_commands.go [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+	accounts, err := r.client.DebugAllAccounts()
+	if err != nil {
+		log.Error("failed to get debug all accounts: %v", err)
+		return
+	}
+
+	for _, a := range accounts {
+		fmt.Println(printPrefix, "Address:", gosmtypes.BytesToAddress(a.AccountId.Address).String())
+		fmt.Println(printPrefix, "Balance:", a.StateCurrent.Balance.Value, coinUnitName)
```
Comment (lrettig) [inline repl/debug_commands.go:20]: I think you might get a null pointer exception here in some cases, be careful. If an account's balance is zero I think `Balance` might be missing. Same for Counter.

File: repl/mock.go [hunk 10]
```diff
@@ -1,6 +1,9 @@
 package repl
 
-import "github.com/spacemeshos/go-spacemesh/accounts"
+import (
+	"github.com/spacemeshos/CLIWallet/common"
+	// types "github.com/spacemeshos/go-spacemesh/common/types"
```
Comment (lrettig) [inline repl/mock.go:5]: undo this change?

File: repl/repl.go [hunk 11]
```diff
@@ -126,258 +168,24 @@ func (r *repl) completer(in prompt.Document) []prompt.Suggest {
...
-	if err != nil {
-		log.Error("failed to list txs: %v", err)
-		return
+	if err := r.client.Sanity(); err != nil {
+		log.Error("Failed to connect to node at %v: %v", r.client.ServerUrl(), err)
+		r.quit()
 	}
 
-	fmt.Println(printPrefix, fmt.Sprintf("txs: %v", txs))
+	fmt.Println("Welcome to Spacemesh. Connected to node at", r.client.ServerUrl())
```
Comment (lrettig) [inline repl/repl.go:186]:
```suggestion
	fmt.Println("Welcome to Spacemesh. Connected to node at: ", r.client.ServerUrl())
```

File: repl/transactions_commands.go [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+	apitypes "github.com/spacemeshos/api/release/go/spacemesh/v1"
+	"github.com/spacemeshos/go-spacemesh/common/util"
+
+	"github.com/spacemeshos/CLIWallet/log"
+	gosmtypes "github.com/spacemeshos/go-spacemesh/common/types"
+)
+
+var transactionStateDisStringsMap = map[int32]string{
+	0: "Unspecified state",
```
Comment (lrettig) [inline repl/transactions_commands.go:16]: Is there a way we can move these strings into the API so that we can use harmonized strings everywhere?

File: repl/transactions_commands.go [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+// helper method - prints tx info
+func printTransaction(t *apitypes.Transaction) {
+
+	txIdStr := "0x" + util.Bytes2Hex(t.Id.Id)
+	fmt.Println(printPrefix, fmt.Sprintf("Transaction id: %v", txIdStr))
+	fmt.Println(printPrefix, "From:", gosmtypes.BytesToAddress(t.Sender.Address).String())
+
+	ct := t.GetCoinTransfer()
+	if ct != nil {
+		fmt.Println(printPrefix, "To (coin account):", gosmtypes.BytesToAddress(ct.Receiver.Address).String())
```
Comment (lrettig) [inline repl/transactions_commands.go:153]:
```suggestion
		fmt.Println(printPrefix, "To (coin account): ", gosmtypes.BytesToAddress(ct.Receiver.Address).String())
```

File: repl/transactions_commands.go [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+		fmt.Println(printPrefix, "Nonce:", t.Counter)
```
Comment (lrettig) [inline repl/transactions_commands.go:154]:
```suggestion
		fmt.Println(printPrefix, "Nonce: ", t.Counter)
```

File: repl/transactions_commands.go [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+		fmt.Println(printPrefix, "Amount:", t.Amount.Value, coinUnitName)
```
Comment (lrettig) [inline repl/transactions_commands.go:155]:
```suggestion
		fmt.Println(printPrefix, "Amount: ", t.Amount.Value, coinUnitName)
```

File: repl/transactions_commands.go [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+		fmt.Println(printPrefix, "Fee:", t.GasOffered.GasProvided, coinUnitName)
```
Comment (lrettig) [inline repl/transactions_commands.go:156]:
```suggestion
		fmt.Println(printPrefix, "Fee: ", t.GasOffered.GasProvided, coinUnitName)
```

**Review by lrettig — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (avive) [reply README.md:59]: It is also there but this is useful. Without it, it is hard to run the CLIWallet for devs who work on it. It wasn't obvious for me..

File: client/mesh_service.go (hunk 4 above)
Comment (avive) [reply client/mesh_service.go:81]: yep, improvement in the future. It is useful to have small requests for small bits of data but maybe also an aggregated one in the api level.

File: repl/transactions_commands.go (hunk 12 above)
Comment (avive) [reply repl/transactions_commands.go:16]: Maybe at a later stage if and when we support localized strings in the api. I don't think it is bad to have these on the client level right now.

File: repl/transactions_commands.go (hunk 13 above)
Comment (avive) [reply repl/transactions_commands.go:153]: No space needed after the colon, it is added automatically between items.

File: README.md (hunk 1 above)
Comment (avive) [reply README.md]: done

File: README.md (hunk 2 above)
Comment (avive) [reply README.md]: done

File: client/node_service.go (hunk 5 above)
Comment (avive) [reply client/node_service.go]: done

File: main.go (hunk 6 above)
Comment (avive) [reply main.go:32]: done

File: repl/account_commands.go (hunk 7 above)
Comment (avive) [reply repl/account_commands.go:66]: done

File: repl/account_commands.go (hunk 8 above)
Comment (avive) [reply repl/account_commands.go]: It is implemented. Please check again.

File: repl/debug_commands.go (hunk 9 above)
Comment (avive) [reply repl/debug_commands.go]: good catch. fixed.

File: repl/mock.go (hunk 10 above)
Comment (avive) [reply repl/mock.go:5]: it is an old leftover. This whole thing needs a major update + more tests. Outside the scope of this pr.

File: repl/repl.go (hunk 11 above)
Comment (avive) [reply repl/repl.go:188]: no need. spaces added automatically between items.

**Comment by avive** [comment]
fixed #17
